// iocluster/index.js
var EventEmitter = require('events').EventEmitter;
var sio = require('socket.io');
var ndata = require('ndata');
var cluster = require('cluster');
var async = require('async');
var handlebars = require('handlebars');

var getGlobalDataKey = function (key) {
	if(key == null) {
		return '__iocl.gld';
	} else {
		return '__iocl.gld.' + key;
	}
};

var getSessionDataKey = function (sessionId, key) {
	if(key == null) {
		return '__iocl.sed' + sessionId;
	} else {
		return '__iocl.sed.' + sessionId + '.' + key;
	}
};

var getSocketDataKey = function (socketId, key) {
	if(key == null) {
		return '__iocl.sod.' + socketId;
	} else {
		return '__iocl.sod.' + socketId + '.' + key;
	}
};

var getGlobalEventKey = function (key) {
	if(key == null) {
		return '__iocl.gle';
	} else {
		return '__iocl.gle.' + key;
	}
};

var getSessionEventKey = function (sessionId, key) {
	if(key == null) {
		return '__iocl.see.' + sessionId;
	} else {
		return '__iocl.see.' + sessionId + '.' + key;
	}
};

var getSocketEventKey = function (socketId, key) {
	if(key == null) {
		return '__iocl.soe.' + socketId;
	} else {
		return '__iocl.soe.' + socketId + '.' + key;
	}
};

var isEmpty = function(obj) {
	var i;
	for(i in obj) {
		return false;
	}
	return true;
};

var AbstractDataClient = function() {

};

AbstractDataClient.prototype.set = function() {
	arguments[0] = this._localizeDataKey(arguments[0]);
	this._dataClient.set.apply(this._dataClient, arguments);
}

AbstractDataClient.prototype.add = function() {
	arguments[0] = this._localizeDataKey(arguments[0]);
	this._dataClient.add.apply(this._dataClient, arguments);
}

AbstractDataClient.prototype.get = function() {
	arguments[0] = this._localizeDataKey(arguments[0]);
	this._dataClient.get.apply(this._dataClient, arguments);
}

AbstractDataClient.prototype.getRange = function() {
	arguments[0] = this._localizeDataKey(arguments[0]);
	this._dataClient.getRange.apply(this._dataClient, arguments);
}

AbstractDataClient.prototype.getAll = function(callback) {
	var clientRootKey = this._localizeDataKey();
	this._dataClient.get.call(this._dataClient, clientRootKey, callback);
}

AbstractDataClient.prototype.count = function() {
	arguments[0] = this._localizeDataKey(arguments[0]);
	this._dataClient.count.apply(this._dataClient, arguments);
}

AbstractDataClient.prototype.remove = function() {
	arguments[0] = this._localizeDataKey(arguments[0]);
	this._dataClient.remove.apply(this._dataClient, arguments);
}

AbstractDataClient.prototype.removeRange = function() {
	arguments[0] = this._localizeDataKey(arguments[0]);
	this._dataClient.removeRange.apply(this._dataClient, arguments);
}

AbstractDataClient.prototype.removeAll = function(callback) {
	var clientRootKey = this._localizeDataKey();
	this._dataClient.set.call(this._dataClient, clientRootKey, {}, callback);
}

AbstractDataClient.prototype.pop = function() {
	arguments[0] = this._localizeDataKey(arguments[0]);
	this._dataClient.pop.apply(this._dataClient, arguments);
}

AbstractDataClient.prototype.hasKey = function() {
	arguments[0] = this._localizeDataKey(arguments[0]);
	this._dataClient.hasKey.apply(this._dataClient, arguments);
}

AbstractDataClient.prototype.stringify = function(value) {
	return this._dataClient.stringify(value);
}

AbstractDataClient.prototype.escape = function(value) {
	return this._dataClient.escape(value);
}

AbstractDataClient.prototype.input = function(value) {
	return this._dataClient.input(value);
}

AbstractDataClient.prototype.extractKeys = function(object) {
	return this._dataClient.extractKeys(object);
}

AbstractDataClient.prototype.extractValues = function(object) {
	return this._dataClient.extractValues(object);
}

AbstractDataClient.prototype.query = function(query, data) {
	var i;
	for(i in data) {
		data[i] = this.input(data[i]);
	}
	var queryTemplate = handlebars.compile(query, {noEscape: true});
	return queryTemplate(data);
}

AbstractDataClient.prototype.run = function(code, callback) {
	this._dataClient.run(code, this._namespace, callback);
}


var Global = function (dataClient, namespace) {
	this._dataClient = dataClient;
	this._namespace = namespace ? namespace : '__main';
};

Global.prototype.__proto__ = AbstractDataClient.prototype;

Global.prototype._localizeDataKey = function (key) {
	return getGlobalDataKey(key);
}

Global.prototype.broadcast = function (event, data, callback) {
	this._dataClient.broadcast(getGlobalEventKey(), {ns: this._namespace, event: event, data: data}, callback);
};

Global.prototype.emit = function (sessionId, event, data, callback) {
	this._dataClient.broadcast(getSessionEventKey(sessionId), {ns: this._namespace, session: sessionId, event: event, data: data}, callback);
};

Global.prototype.ns = function (namespace) {
	return new Global(this._dataClient, namespace);
};


var Session = function (sessionId, dataClient, namespace) {
	this.id = sessionId;
	this._dataClient = dataClient;
	this._namespace = namespace ? namespace : '__main';
};

Session.prototype.__proto__ = AbstractDataClient.prototype;

Session.prototype._localizeDataKey = function (key) {
	return getSessionDataKey(this.id, key);
}

Session.prototype.setAuth = function(data, callback) {
	this.set('__auth', data, callback);
}

Session.prototype.getAuth = function(callback) {
	this.get('__auth', callback);
}

Session.prototype.clearAuth = function(callback) {
	this.remove('__auth', callback);
}

Session.prototype.emit = function (event, data, callback) {
	this._dataClient.broadcast(getSessionEventKey(this.id), {ns: this._namespace, session: this.id, event: event, data: data}, callback);
};

Session.prototype.ns = function (namespace) {
	return new Session(this.id, this._dataClient, namespace);
};


var Socket = function (socketId, dataClient, namespace) {
	this.id = socketId;
	this._dataClient = dataClient;
	this._namespace = namespace ? namespace : '__main';
};

Socket.prototype.__proto__ = AbstractDataClient.prototype;

Socket.prototype._localizeDataKey = function (key) {
	return getSocketDataKey(this.id, key);
}

Socket.prototype.emit = function (event, data, callback) {
	this._dataClient.broadcast(getSocketEventKey(this.id), {ns: this._namespace, socket: this.id, event: event, data: data}, callback);
}

Socket.prototype.ns = function (namespace) {
	return new Socket(this.id, this._dataClient, namespace);
};


var IOClusterServer = module.exports.IOClusterServer = function (port, secretKey) {
	var self = this;
	this._dataServer = ndata.createServer(port, secretKey);
	
	this._dataServer.on('ready', function() {
		self.emit('ready');
	});
};

IOClusterServer.prototype.__proto__ = EventEmitter.prototype;


var IOClusterClient = module.exports.IOClusterClient = function (port, secretKey) {
	var self = this;
	
	this._dataClient = ndata.createClient(port, secretKey);
	
	this._dataClient.on('ready', function () {
		self._dataClient.watchOnce(getGlobalEventKey(), function () { self._handleGlobalEvent.apply(self, arguments); }, function(err) {
			if(err) {
				self.emit('fail', err);
			} else {
				self.emit('ready');
			}
		});
	});
	
	this._sockets = {};
	this._sessions = {};
};

IOClusterClient.prototype.__proto__ = EventEmitter.prototype;

IOClusterClient.prototype.bind = function (socket, callback) {
	var self = this;
	this._sockets[socket.id] = socket;
	if(!this._sessions.hasOwnProperty(socket.ssid)) {
		this._sessions[socket.ssid] = {};
	}
	this._sessions[socket.ssid][socket.id] = socket;
	
	async.parallel([
		function (cb) {
			self._dataClient.watchOnce(getSocketEventKey(socket.id), function () { self._handleSocketEvent.apply(self, arguments); }, cb);
		},
		function (cb) {
			self._dataClient.watchOnce(getSessionEventKey(socket.ssid), function () { self._handleSessionEvent.apply(self, arguments); }, cb);
		},
		function (cb) {
			self._dataClient.set('__iocl.sessions.' + self._dataClient.escape(socket.ssid) + '.' + self._dataClient.escape(socket.id), 1, cb);
		},
		function (cb) {
			self._dataClient.set('__iocl.addresses.' + socket.address + '.' + self._dataClient.escape(socket.id), 1, cb);
		},
	],
	function(err) {
		callback && callback(err, socket);
	});
};

IOClusterClient.prototype.unbind = function (socket, callback) {
	var self = this;
	
	this._dataClient.unwatch(getSocketEventKey(socket.id), null, function (err) {
		if(err) {
			callback(err);
		} else {
			delete self._sockets[socket.id];
			delete self._sessions[socket.ssid][socket.id];
			
			if(isEmpty(self._sessions[socket.ssid])) {
				delete self._sessions[socket.ssid];
				
				async.waterfall([
					function (cb) {
						self._dataClient.unwatch(getSessionEventKey(socket.ssid), null, cb);
					},
					function (cb) {
						self._dataClient.remove('__iocl.sessions.' + self._dataClient.escape(socket.ssid) + '.' + self._dataClient.escape(socket.id), cb);
					},
					function (cb) {
						self._dataClient.count('__iocl.sessions.' + self._dataClient.escape(socket.ssid), function (err, value) {
							if(!err && value < 1) {
								self._dataClient.remove('__iocl.sessions.' + self._dataClient.escape(socket.ssid), function (err) {
									if(err) {
										cb(err);
									} else {
										self._dataClient.remove('__iocl.addresses.' + socket.address, cb);
									}
								});
							} else {
								cb(err);
							}
						});
					}
				], callback);
			}
		}
	});
};

IOClusterClient.prototype.getAddressSockets = function (ipAddress, callback) {
	self._dataClient.get('__iocl.addresses.' + self._dataClient.escape(ipAddress), function (err, data) {
		var sockets = [];
		var i;
		for(i in data) {
			sockets.push(i);
		}
		callback(err, sockets);
	});
};

IOClusterClient.prototype.global = function () {
	return new Global(this._dataClient);
};

IOClusterClient.prototype.session = function (sessionId) {
	return new Session(sessionId, this._dataClient);
};

IOClusterClient.prototype.socket = function (socketId) {
	return new Socket(socketId, this._dataClient);
};

IOClusterClient.prototype._handleSocketEvent = function (e) {
	var eventObject = {
		ns: e.ns,
		event: e.event,
		data: e.data
	};
	
	if(this._sockets.hasOwnProperty(e.socket)) {
		this._sockets[e.socket].emit('event', eventObject);
	}
};

IOClusterClient.prototype._handleSessionEvent = function (e) {
	var eventObject = {
		ns: e.ns,
		event: e.event,
		data: e.data
	};
	
	if(this._sessions.hasOwnProperty(e.session)) {
		var sockets = this._sessions[e.session];
		var i;
		for(i in sockets) {
			sockets[i].emit('event', eventObject);
		}
	}
};

IOClusterClient.prototype._handleGlobalEvent = function (e) {
	var i;
	var eventObject = {
		ns: e.ns,
		event: e.event,
		data: e.data
	};
	
	for(i in this._sockets) {
		this._sockets[i].emit('event', eventObject);
	}
};
