// expirymanager/index.js
var ExpiryManager = module.exports.ExpiryManager = function () {
	this._keys = {};
    this._expiries = {};
};

ExpiryManager.prototype._isEmpty = function (obj) {
	var i;
	for (i in obj) {
		return false;
	}
	return true;
};

ExpiryManager.prototype.now = function () {
	return (new Date()).getTime() / 1000;
};

ExpiryManager.prototype.expire = function (key, seconds) {
	this.unexpire(key);
	var expiry = this.now() + seconds;
    this._keys[key] = expiry;
	if (this._expiries[expiry] == null) {
		this._expiries[expiry] = {};
	}
	this._expiries[expiry][key] = 1;
};

ExpiryManager.prototype.getExpiry = function (key) {
	return this._keys[key];
};

ExpiryManager.prototype.getKeysByExpiry = function (expiry) {
	var keys = [];
	var keyMap = this._expiries[expiry];
	var i;
	for(i in keyMap) {
		keys.push(i);
	}
	return keys;
};

ExpiryManager.prototype.unexpire = function (key) {
	var expiry = this._keys[key];
	delete this._keys[key];
	if (expiry && this._expiries[expiry] != null) {
		delete this._expiries[expiry][key];
		if (this._isEmpty(this._expiries[expiry])) {
			delete this._expiries[expiry];
		}
	}
};

ExpiryManager.prototype.getExpiredKeys = function (time) {
	var expiredKeys = [];
	var now = time || this.now();
	var i, j;
	for (i in this._expiries) {
		if (i <= now) {
			for (j in this._expiries[i]) {
				expiredKeys.push(j);
			}
		} else {
			break;
		}
	}
	return expiredKeys;
};

ExpiryManager.prototype.extractExpiredKeys = function (time) {
	var expiredKeys = this.getExpiredKeys(time);
	var i;
	for (i in expiredKeys) {
		this.unexpire(expiredKeys[i]);
	}
	return expiredKeys;
};

ExpiryManager.prototype.clear = function () {
	this._keys = {};
	this._expiries = {};
};
